pkg/configs/db/postgres/postgres.go

```diff
@@ -134,6 +135,86 @@ func (d DB) GetConfigs(since configs.ID) (map[string]configs.View, error) {
 	})
 }
 
+// GetRulesConfig gets the latest alertmanager config for a user.
+func (d DB) GetRulesConfig(userID string) (configs.VersionedRulesConfig, error) {
+	current, err := d.GetConfig(userID)
+	if err != nil {
+		return configs.VersionedRulesConfig{}, err
+	}
+	cfg := current.GetVersionedRulesConfig()
+	if cfg == nil {
+		return configs.VersionedRulesConfig{}, sql.ErrNoRows
+	}
+	return *cfg, nil
+}
+
+// SetRulesConfig sets the current alertmanager config for a user.
+func (d DB) SetRulesConfig(userID string, oldConfig, newConfig configs.RulesConfig) (bool, error) {
+	updated := false
+	err := d.Transaction(func(tx DB) error {
+		current, err := d.GetConfig(userID)
+		if err != nil && err != sql.ErrNoRows {
+			return err
+		}
+		// The supplied oldConfig must match the current config. If no config
+		// exists, then oldConfig must be nil. Otherwise, it must exactly
+		// equal the existing config.
+		if !((err == sql.ErrNoRows && oldConfig == nil) || reflect.DeepEqual(current.Config.RulesFiles, oldConfig)) {
+			return nil
+		}
+		new := configs.Config{
+			AlertmanagerConfig: current.Config.AlertmanagerConfig,
+			RulesFiles:         newConfig,
+		}
+		updated = true
+		return d.SetConfig(userID, new)
+	})
+	return updated, err
+}
+
+func (d DB) findRulesConfigs(filter squirrel.Sqlizer) (map[string]configs.VersionedRulesConfig, error) {
+	rows, err := d.Select("id", "owner_id", "config ->> 'rules_files'").
+		Options("DISTINCT ON (owner_id)").
+		From("configs").
+		Where(filter).
+		Where("config ->> 'rules_files' <> '{}'").
```

Comment:
Why?

Reply:
Good question. I've added a fairly detailed comment, repasted here in case that assists your follow-up review.

```go
		// `->>` gets a JSON object field as text. When a config row exists
		// and alertmanager config is provided but ruler config has not yet
		// been, the 'rules_files' key will have an empty JSON object as its
		// value. This is (probably) the most efficient way to test for a
		// non-empty `rules_files` key.
		//
		// This whole situation is way too complicated. See
		// https://github.com/weaveworks/cortex/issues/619 for the whole
		// story, and our plans to improve it.
```

Reply:
This seems to assume we only ever go from blank to non-blank in the history; if it goes from non-blank to blank then you will select the previous version.